[DXIL][Analysis] Add DXILMetadataAnalysis pass

bharadwajy opened the pull request:
DXIL Metadata Analysis passes (one for legacy PM and one for new PM) that collect following DXIL module metadata information in a structure are added. 
1. Shader Model version
2. DXIL version 
3. Shader Stage

Information collected using the legacy pass is verified by adding additional test commands to existing metadata test sources.

Commit:
[DXIL][Analysis] Implementation of DXILMetadataAnalysis pass
Three peices of information viz., Shader Model, DXIL version and Shader
Stage information are collected in a structure. Additional metadata
information is planned to be added in subsequent PRs along with changes
to consume the information in passes such as DXILTranslateMetadata.

python3kgae commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -0,0 +1,83 @@
+
+/// Wrapper pass to be used by other passes using legacy pass manager
+class DXILMetadataAnalysisWrapperPass : public ModulePass {
+  std::unique_ptr<dxil::ModuleMetadataInfo> ModuleMetadata;
```
Could this just be a normal variable?
```
dxil::ModuleMetadataInfo ModuleMetadata
```
Then we don't need the releaseMemory.

bharadwajy replied:
> Could this just be a normal variable?
> 
> ```
> dxil::ModuleMetadataInfo ModuleMetadata
> ```
> 
> Then we don't need the releaseMemory.

Changed.

bogner replied:
This just means that we never release the memory though? While the `dxil::ModuleMetadataInfo` is (currently) only about 6.5x the size of a pointer it's likely to continue growing, and the [lifetime of a pass in the legacy pass manager is the whole compilation process](https://llvm.org/docs/WritingAnLLVMPass.html#the-releasememory-method). It really is better to just use a unique_ptr here.

bharadwajy replied:
> This just means that we never release the memory though? While the `dxil::ModuleMetadataInfo` is (currently) only about 6.5x the size of a pointer it's likely to continue growing, and the [lifetime of a pass in the legacy pass manager is the whole compilation process](https://llvm.org/docs/WritingAnLLVMPass.html#the-releasememory-method). It really is better to just use a unique_ptr here.

Changed to use `std::unique_ptr`.

bogner commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -0,0 +1,93 @@
+llvm::dxil::ModuleMetadataInfo
+DXILMetadataAnalysis::run(Module &M, ModuleAnalysisManager &AM) {
+  ModuleMetadataInfo Data;
+  return Data;
```
This just creates an empty `ModuleMetadataInfo` without actually populating it

bharadwajy replied:
> This just creates an empty `ModuleMetadataInfo` without actually populating it

Fixed.

bogner commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -0,0 +1,93 @@
+  Triple TT(Triple(M.getTargetTriple()));
+  ModuleMetadata->DXILVersion = TT.getDXILVersion();
+  ModuleMetadata->ShaderModel = TT.getOS();
+  ModuleMetadata->ShaderStage = TT.getEnvironment();
```
We shouldn't put the logic for the old PM pass in the runOnModule method itself. We need to break out a helper that can be used by both the old and new PM.

bharadwajy replied:
> We shouldn't put the logic for the old PM pass in the runOnModule method itself. We need to break out a helper that can be used by both the old and new PM.

Changed.

bogner reviewed (commented):
The title calls this boilerplate, but it does actually partially implement the actual logic. We should drop "boilerplate" from the title and add tests for this please.

python3kgae commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -0,0 +1,83 @@
+namespace dxil {
+
+struct ModuleMetadataInfo {
+  VersionTuple DXILVersion;
```
We'll need ValidatorVersion as well.

bharadwajy replied:
> We'll need ValidatorVersion as well.

Added.

bharadwajy commented:
> The title calls this boilerplate, but it does actually partially implement the actual logic. We should drop "boilerplate" from the title and add tests for this please.

Updated title and description. Added tests.

bogner commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -0,0 +1,85 @@
+  VersionTuple ShaderModelVersion{};
+  Triple::EnvironmentType ShaderStage = Triple::UnknownEnvironment;
+
+  void init(Module &);
```
Rather than having an `init` method in the public interface like this it's better to construct the analysis result in a `static` function or a helper class in an anonymous namespace in the `.cpp` file. That way we don't need to leak implementation details to the header.

bogner replied:
A constructor is still in the public interface, and the idea is really that we want the initialization to be internal to the analysis. It doesn’t make sense to create this object outside of the analysis, as the analysis is the way to get the analysis result.

Further, if the helper that builds this is static/anonymous, then if we need to make it depend on other analyses or something like that the changes can be contained in the cpp file, which is nice for encapsulation.

bharadwajy replied:
Changes pushed to use an anonymously scoped function instead of using a constructor.

bogner replied:
Style nit: llvm usually only uses anonymous namespaces for classes/structs. Local functions are generally just declared “static”.

bharadwajy replied:
> Style nit: llvm usually only uses anonymous namespaces for classes/structs. Local functions are generally just declared “static”.

Changed to `static` function.

bogner commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -0,0 +1,85 @@
+  Triple::EnvironmentType ShaderStage = Triple::UnknownEnvironment;
+
+  void init(Module &);
+  void dump(raw_ostream &OS = dbgs()) const;
```
This is not a dump method (in the `LLVMP_DUMP_METHOD` / `LLVM_ENABLE_DUMP` sense) and shouldn't be called as such. Better to just have a `void print(raw_ostream &OS) const` and not default to `dbgs()`.

bogner commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -0,0 +1,85 @@
+};
+
+/// Legacy pass to be used by other passes using legacy pass manager
+class DXILMetadataAnalysisLegacyPass : public ModulePass {
```
I'm not sure how consistent the naming conventions are here so maybe it doesn't matter too much, but I think the legacy PM versions of analysis passes are usually named as "wrapper" not "legacy", as in they are legacy wrappers for the new PM pass.

bogner commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -0,0 +1,85 @@
+
+/// Legacy pass to be used by other passes using legacy pass manager
+class DXILMetadataAnalysisLegacyPass : public ModulePass {
+  dxil::ModuleMetadataInfo Data;
+  bool AnalysisDone = false;
```
As mentioned in my reply to the conversation with @python3kgae, I think we should go back to a unique_ptr here.

bogner commented on llvm/lib/Target/DirectX/DirectX.h:
```diff
@@ -58,6 +58,12 @@ void initializeDXILPrettyPrinterPass(PassRegistry &);
 /// Initializer for dxil::ShaderFlagsAnalysisWrapper pass.
 void initializeShaderFlagsAnalysisWrapperPass(PassRegistry &);
 
+/// Initializer for dxil::DXILMetadataAnalysisLegacy pass.
+void initializeDXILMetadataAnalysisLegacyPassPass(PassRegistry &);
+
+/// Initializer for dxil::DXILMetadataAnalysisLegacyPrinter pass.
+void initializeDXILMetadataAnalysisLegacyPrinterPass(PassRegistry &);
+
```
These are declared in `include/llvm/InitializePasses.h`, they shouldn't be here as well.

bogner commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -0,0 +1,151 @@
+// Pass to print - for verification - of analysis information collected in
+// DXILMetadataAnalysisLegacyPass
+
+class DXILMetadataAnalysisLegacyPrinter : public ModulePass {
+public:
+  static char ID;
+
+  DXILMetadataAnalysisLegacyPrinter();
+
+  bool runOnModule(Module &M) override;
+  void getAnalysisUsage(AnalysisUsage &AU) const override;
+};
```
I don't know that we need to bother with the legacy printer pass - we can just use the NewPM printer pass in the tests (that is, `opt -passes=print<dxil-metadata>` instead of `-dxil-metadata-analysis-print`)

bogner reviewed (commented):
This is still missing changes in `llvm/lib/Passes/PassRegistry.def` to add the newpm style passes to the pass manager. Take a look at #100700 if you want some code to follow for the pattern.

bogner commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -0,0 +1,106 @@
+  DXILVersion = TT.getDXILVersion();
+  ShaderModelVersion = TT.getOSVersion();
+  ShaderStage = TT.getEnvironment();
+  NamedMDNode *Entry = M.getOrInsertNamedMetadata("dx.valver");
```
Analysis passes must not modify the module, so creating metadata here isn’t allowed. In any case, we should be creating the `dx.valver` metadata node based on the results of this pass, so getting the value from there is circular and doesn’t make sense.

bharadwajy replied:
> Analysis passes must not modify the module, so creating metadata here isn’t allowed. In any case, we should be creating the `dx.valver` metadata node based on the results of this pass, so getting the value from there is circular and doesn’t make sense.

Changed the call to get named metadata `dx.ver` from the `Module` along with a change to set `ValidatorVersion` only if one exists, else retaining its default empty value. Is there a different location to obtain Validator Version?

bogner replied:
My understanding is that validator version is derived from the shader model / DXIL version, and that the only time that it isn’t is if it’s overridden on the command line. There was some discussion recently about why you would ever want to override it other than specifically for preview versions and what it even means for the validator version to differ significantly from the shader model version. Currently the metadata you’re reading is only ever set by the command line override.

Presumably the “right” thing to do here is to derive the validator version from the other versions. However, if we want to maintain the current behaviour in clang it looks like it just defaults to 1.0 if the command line argument isn’t set (this seems wrong). I think there’s a valid argument to leaving this out of this change and having a discussion about what the correct behaviour here actually is before adding it.

bharadwajy replied:
...
> ...I think there’s a valid argument to leaving this out of this change and having a discussion about what the correct behaviour here actually is before adding it.

Excluded validator version information from `struct ModuleMetadataInfo`
